File: frontend/src/lib/exonSkipping.ts
export interface EseRegion {
  start: number;
  end: number;
  a_positions: number[];
  matrices: string[];
  peak_score: number;
}

export interface ExonSkippingCandidate {
  id: string;
  kind: "sa" | "ese";
  label: string;
  target_sequence: string;
  arrna_sequence: string;
  global_start: number;
  global_end: number;
  exon_start: number;
  exon_end: number;
  mismatch_target_position: number | null;
  mismatch_arrna_position: number | null;
  covered_a_positions: number[];
  covered_region_indices: number[];
}

export interface ExonSkippingDesign {
  schema_version: "1.0";
  application: "exon_skipping";
  normalized_input: {
    upstream_intron: string;
    exon: string;
    downstream_intron: string;
    combined_length: number;
    coordinate_system: "concatenated_sequence_1_based";
  };
  normalized_upstream_intron: string;
  normalized_exon: string;
  normalized_downstream_intron: string;
  arrna_length: number;
  exon_length: number;
  length_class: "short" | "medium" | "long";
  sa_is_canonical: boolean;
  warnings: string[];
  ese_hits: unknown[];
  ese_regions: EseRegion[];
  candidates: ExonSkippingCandidate[];
  transcript: string | null;
  exon_number: number | null;
  provenance: import("./api").DesignProvenance;
}

export interface NcbiExonReference {
  species: string;
  gene: string;
  transcript: string;
  transcript_name: string;
  selection: "MANE Select" | "longest protein-coding";
  exon_number: number;
  exon_count: number;
  genomic_accession: string;
  orientation: "plus" | "minus";
  upstream_intron: string;
  exon: string;
  downstream_intron: string;
}

export interface NcbiCodingReference {
  species: string;
  gene: string;
  transcript: string;
  transcript_name: string;
  selection: "MANE Select" | "longest protein-coding";
  cds_sequence: string;
  cds_start: number;
  cds_end: number;
}

import { apiPost } from "./api";

export function resolveNcbiExon(species: string, gene: string, exonNumber: number, flankLength: number) {
  return apiPost<NcbiExonReference>("/references/ncbi/exon", {
    species, gene, exon_number: exonNumber, flank_length: flankLength,
  });
}

export function resolveNcbiCodingSequence(species: string, gene: string) {
  return apiPost<NcbiCodingReference>("/references/ncbi/coding-sequence", { species, gene });
}

export function createExonSkippingDesign(
  upstreamIntron: string,
  exon: string,
  downstreamIntron: string,
  arrnaLength: number,
  reference?: NcbiExonReference,
) {
  return apiPost<ExonSkippingDesign>("/exon-skipping/design", {
    upstream_intron: upstreamIntron,
    exon,
    downstream_intron: downstreamIntron,
    arrna_length: arrnaLength,
    transcript: reference?.transcript,
    exon_number: reference?.exon_number,
  });
}

File: backend/app/schemas/exon_skipping.py
from typing import Literal

from pydantic import BaseModel, Field, field_validator

from app.schemas.design import DesignProvenance


class ExonSkippingDesignRequest(BaseModel):
    upstream_intron: str = Field(min_length=1)
    exon: str = Field(min_length=1)
    downstream_intron: str = Field(min_length=1)
    arrna_length: int = Field(default=151, ge=21, le=1001)
    transcript: str | None = None
    exon_number: int | None = None

    @field_validator("upstream_intron", "exon", "downstream_intron")
    @classmethod
    def validate_sequence(cls, value: str) -> str:
        sequence = "".join(value.upper().split()).replace("U", "T")
        if not sequence or set(sequence) - set("ACGT"):
            raise ValueError("Only A, C, G, T or U are supported")
        return sequence


class EseHit(BaseModel):
    matrix: str
    score: float
    start: int
    end: int
    sequence: str
    a_positions: list[int]


class EseRegion(BaseModel):
    start: int
    end: int
    a_positions: list[int]
    matrices: list[str]
    peak_score: float


class ArrnaCandidate(BaseModel):
    id: str
    kind: Literal["sa", "ese"]
    label: str
    target_sequence: str
    arrna_sequence: str
    global_start: int
    global_end: int
    exon_start: int
    exon_end: int
    mismatch_target_position: int | None = None
    mismatch_arrna_position: int | None = None
    covered_a_positions: list[int] = []
    covered_region_indices: list[int] = []


class ExonSkippingNormalizedInput(BaseModel):
    upstream_intron: str
    exon: str
    downstream_intron: str
    combined_length: int
    coordinate_system: Literal["concatenated_sequence_1_based"] = (
        "concatenated_sequence_1_based"
    )


class ExonSkippingDesignResponse(BaseModel):
    schema_version: Literal["1.0"] = "1.0"
    application: Literal["exon_skipping"] = "exon_skipping"
    normalized_input: ExonSkippingNormalizedInput
    normalized_upstream_intron: str
    normalized_exon: str
    normalized_downstream_intron: str
    arrna_length: int
    exon_length: int
    length_class: Literal["short", "medium", "long"]
    sa_is_canonical: bool
    warnings: list[str]
    ese_hits: list[EseHit]
    ese_regions: list[EseRegion]
    candidates: list[ArrnaCandidate]
    transcript: str | None = None
    exon_number: int | None = None
    provenance: DesignProvenance


class NcbiExonRequest(BaseModel):
    species: str = Field(min_length=1)
    gene: str = Field(min_length=1)
    exon_number: int = Field(ge=2)
    flank_length: int = Field(default=151, ge=21, le=1001)


class NcbiExonResponse(BaseModel):
    species: str
    gene: str
    transcript: str
    transcript_name: str
    selection: Literal["MANE Select", "longest protein-coding"]
    exon_number: int
    exon_count: int
    genomic_accession: str
    orientation: Literal["plus", "minus"]
    upstream_intron: str
    exon: str
    downstream_intron: str


class NcbiCodingSequenceRequest(BaseModel):
    species: str = Field(min_length=1)
    gene: str = Field(min_length=1)


class NcbiCodingSequenceResponse(BaseModel):
    species: str
    gene: str
    transcript: str
    transcript_name: str
    selection: Literal["MANE Select", "longest protein-coding"]
    cds_sequence: str
    cds_start: int
    cds_end: int
